### A number spiral
(Source: https://medium.com/mathematica-stories/why-is-2017-an-interesting-number-dc3cfc135853)

Arrange the odd numbers in a square spiral, as shown below. If you start at the 1 in the center and move down 16 cells then you will land on the number 2017. Neat, huh? (How this is related to the lazy caterer’s sequence?)

<img src = 'https://cdn-images-1.medium.com/max/800/1*MpvMlo2PA2UhWNpVpEMCxA.png' />

- The code below helps us verify that the number 16 cells down is indeed 2017.
- This code allows us to create a clockwise spiral where we can specify:
 - the starting number (start_num)
 - the step size (step_size) and 
 - the number of loops in the spiral (num_loops)
 
The relationship to the [lazy caterer sequence](https://en.wikipedia.org/wiki/Lazy_caterer%27s_sequence) is explained below.

If you have not done so already, check out my [lazy caterer program](https://github.com/KobyO/Misc/blob/master/Lazy%20Caterer%20Problem.ipynb).

In [1]:
import numpy as np

In [2]:
class Spiral:
    
    spiral = {} # dictionary to represent the spiral
    cardinals = {'east': 1, 'south': 3, 'west': 5, 'north': 7}
    
    start_num = 0
    step_size = 0
    num_loops = 0
    
    def __init__(self, start_num, step_size, num_loops):
        self.spiral = {}
        self.start_num = start_num
        self.step_size = step_size
        self.num_loops = num_loops
        
        #initialize keys:
        for loop_index in range(num_loops + 1): #spiral[0] is not counted as a loop, hence the +1
            self.spiral[loop_index] = []

        #initialize starting point:
        self.spiral[0] = start_num
        next_item = start_num + step_size

        #populate the spiral
        for loop_index in range(1, num_loops + 1):
            for loop_slot in range(8*loop_index):
                self.spiral[loop_index].append(next_item)
                next_item += step_size
        
    
    def get_start_num(self):
        return self.start_num
    
    def get_step_size(self):
        return self.step_size
    
    def get_num_loops(self):
        return self.num_loops
    
    def read(self, direction): # read the spiral in a specified direction

        output = []
        output.append(self.spiral[0])
        step = self.cardinals[direction] - 1 # first index in 'direction'

        for loop in range(1, len(self.spiral)):
            output.append(self.spiral[loop][step])
            step += self.cardinals[direction]

        return output
    
    def display(self):
        
        grid_size = 2*self.num_loops + 1
        center = grid_size//2

        spiral_matrix = np.zeros((grid_size, grid_size)).astype(np.int)
        spiral_matrix[center][center] = self.spiral[0] #initialize center with start_num

        top_row = 0
        left_col = 0
        right_col = grid_size - 1
        bottom_row = grid_size - 1

        for loop_num in reversed(range(1, self.num_loops + 1)): # start from outermost loop

            loop_length = len(self.spiral[loop_num])
            row_size = int(loop_length/4) + 1
            col_size = int(loop_length/4) - 1
            row_start = left_col
            col_start = top_row + 1

            spiral_matrix[top_row, row_start:(row_start + row_size)] = self.spiral[loop_num][-(row_size):]
            spiral_matrix[bottom_row, row_start:(row_start + row_size)] = list(reversed(self.spiral[loop_num][col_size:(col_size + row_size)]))
            spiral_matrix[col_start:(col_start + col_size), left_col] = list(reversed(self.spiral[loop_num][-(col_size + row_size):-(row_size)]))
            spiral_matrix[col_start:(col_start + col_size), right_col] = self.spiral[loop_num][:col_size]

            top_row += 1
            left_col += 1
            right_col -= 1
            bottom_row -= 1

        return spiral_matrix
    
    def scramble(self):
        return

In [3]:
# Create a spiral with start_num = 1, step_size = 2, num_loops = 16
s = Spiral(1, 2, 16)

In [4]:
# We can read the spiral in the south direction to see that the last number is 2017
print(s.read('south'))

[1, 7, 29, 67, 121, 191, 277, 379, 497, 631, 781, 947, 1129, 1327, 1541, 1771, 2017]


- By the way, it turns out that the spiral is related to the Lazy Caterer problem as follows:
  
  If n is the value located c cells the south (i.e. downward),
  then n is the number of pieces produced by 4c - 1 straight cuts in the Lazy Caterer problem.
  
  For example, the 14th value in s.read('south') is 1541, which is the number of pieces produced by 4(14) - 1 straight cuts (i.e. 55 straight cuts) in the Lazy Caterer problem).

In [5]:
# The entire spiral is too large to display all the values
s.display()

array([[2113, 2115, 2117, ..., 2173, 2175, 2177],
       [2111, 1861, 1863, ..., 1919, 1921, 1923],
       [2109, 1859, 1625, ..., 1681, 1683, 1925],
       ...,
       [2053, 1803, 1569, ..., 1513, 1739, 1981],
       [2051, 1801, 1799, ..., 1743, 1741, 1983],
       [2049, 2047, 2045, ..., 1989, 1987, 1985]])

In [6]:
# But just for fun, we can create a smaller spiral and display it
s_small = Spiral(1, 2, 5)
s_small.display()

array([[221, 223, 225, 227, 229, 231, 233, 235, 237, 239, 241],
       [219, 145, 147, 149, 151, 153, 155, 157, 159, 161, 163],
       [217, 143,  85,  87,  89,  91,  93,  95,  97,  99, 165],
       [215, 141,  83,  41,  43,  45,  47,  49,  51, 101, 167],
       [213, 139,  81,  39,  13,  15,  17,  19,  53, 103, 169],
       [211, 137,  79,  37,  11,   1,   3,  21,  55, 105, 171],
       [209, 135,  77,  35,   9,   7,   5,  23,  57, 107, 173],
       [207, 133,  75,  33,  31,  29,  27,  25,  59, 109, 175],
       [205, 131,  73,  71,  69,  67,  65,  63,  61, 111, 177],
       [203, 129, 127, 125, 123, 121, 119, 117, 115, 113, 179],
       [201, 199, 197, 195, 193, 191, 189, 187, 185, 183, 181]])